### PR TITLE
Pulse Compiler - `BroadcastInstructions` pass

### PR DESCRIPTION
**Pull request by TsafrirA**
### Summary
A `BroadcastInstructions` pass is added, to convert `Frame` \ `PulseTarget` only instructions into a set of the relevant `MixedFrame`s instructions (while keeping sequencing and timing intact).

### CODE REVIEW

File: qiskit/pulse/compiler/passes/broadcast_instructions.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+from __future__ import annotations
+
+from qiskit.pulse.compiler.basepasses import TransformationPass
+from qiskit.pulse.ir import SequenceIR
+from qiskit.pulse import Frame, PulseTarget
+from qiskit.pulse.exceptions import PulseCompilerError
+
+
+class BroadcastInstructions(TransformationPass):
+    r"""Broadcast :class:`~qiskit.pulse.Frame` or :class:`~qiskit.pulse.PulseTarget` only instructions.
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/broadcast_instructions.py:24]:
```suggestion
    """Broadcast :class:`~qiskit.pulse.Frame` or :class:`~qiskit.pulse.PulseTarget` only instructions.
```
Do I miss any math syntax?

File: qiskit/pulse/compiler/passes/broadcast_instructions.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            elif isinstance(inst_target := elem.inst_target, (Frame, PulseTarget)):
+                in_edges = [x[0] for x in prog.sequence.in_edges(ind)]
+                out_edges = [x[1] for x in prog.sequence.out_edges(ind)]
+                initial_time = prog.time_table[ind]
+                if mixed_frames := mixed_frames_mapping[inst_target]:
+                    for mixed_frame in mixed_frames:
+                        # The relevant instructions are delay and set\shift phase\frequency, and they all
+                        # have the same signature.
+                        new_ind = prog.sequence.add_node(
+                            elem.__class__(
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/broadcast_instructions.py:100]: Seems like this logic is fragile. I think this is the situation where MixIn class fits in; namely

```python
class PartialMixFrameMixIn:
    def bind_mix_frame(self, mixed_frame) -> Instruction:
        # return new instance with given mixed frame

class FrameUpdate(Instruction, PartialMixFrameMixIn)
    ...

class Delay(Instruction, PartialMixFrameMixIn)
    ...
```

and then you can write

```python
if isinstance(elem, PartialMixFrameMixIn):  # thanks to mro
    new_elem = elem.bind_mix_frame(mixed_frame)
```

File: qiskit/pulse/compiler/passes/broadcast_instructions.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                                elem.operands[0], mixed_frame=mixed_frame, name=elem.name
+                            )
+                        )
+                        prog.sequence.add_edges_from_no_data(
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/broadcast_instructions.py:104]: Do we need all edges for this node with a particular mixed frame? I think we can also create an edge for the in-node with the same mixed frame (same for out edge). Probably this depends how we want to design the semantic -- if we want to treat the frame update in the same way with barrier this implementation is correct. Alternatively we can also ask user to explicitly write barrier if they want to block subsequent instructions.

File: qiskit/pulse/compiler/passes/broadcast_instructions.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    ) -> SequenceIR:
+        """Run broadcasting pass.
+
+        Arguments:
+            passmanager_ir: The IR object to undergo broadcasting.
+
+        Raises:
+            PulseCompilerError: if ``property_set`` does not include a mixed_frames_mapping dictionary.
+        """
+        if self.property_set["mixed_frames_mapping"] is None:
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/broadcast_instructions.py:68]: Can you also check the edges of input graph and raise a user warning if the edge length is zero?

**Review by nkanazawa1989 — COMMENTED** [review]
Thanks. One suggestion and some minor comments.

File: qiskit/pulse/compiler/passes/broadcast_instructions.py (hunk 1 above)
Comment (TsafrirA) [reply qiskit/pulse/compiler/passes/broadcast_instructions.py:24]: I think it's needed for `\s`, though I always get this wrong.

**Comment by TsafrirA** [comment]
Closing the PR because development has stopped, and Pulse is being removed from Qiskit.

**Pull request closed without merging**